go/pkg/cas/upload.go

```diff
@@ -39,6 +65,361 @@ type DigestStat struct {
 
 // Upload uploads all inputs. It exits when inputC is closed or ctx is canceled.
 func (c *Client) Upload(ctx context.Context, inputC <-chan *UploadInput) (stats *TransferStats, err error) {
+	eg, ctx := errgroup.WithContext(ctx)
+	// Do not exit until all sub-goroutines exit, to prevent goroutine leaks.
+	defer eg.Wait()
+
+	u := &uploader{
+		Client: c,
+		eg:     eg,
+
+		fsSem: semaphore.NewWeighted(int64(c.FSConcurrency)),
+	}
+
+	// Start processing input.
+	eg.Go(func() error {
+		for {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case in, ok := <-inputC:
+				if !ok {
+					return nil
+				}
+				u.eg.Go(func() error {
+					return errors.Wrapf(u.startProcessing(ctx, in), "%q", in.Path)
+				})
+			}
+		}
+	})
+	return &u.stats, eg.Wait()
+}
+
+// uploader implements a concurrent multi-stage pipeline to read blobs from the
+// file system, check their presence on the server and then upload if necessary.
+// Common blobs are deduplicated.
+//
+// uploader.eg is used to schedule work, while concurrency of individual
+// expensive operations is controlled via separate semaphores.
+//
+// Special care is taken for large files: they are read sequentially, opened
+// only once per file, and read with large IO size.
+type uploader struct {
+	*Client
+	eg    *errgroup.Group
+	stats TransferStats
+
+	// already-processed files.
+	// Types fsCacheKey and fsCacheValue are used for keys/values respectively.
+	fsCache sync.Map
+	// controls concurrency of file IO.
+	// TODO(nodir): ensure it does not hurt streaming.
+	fsSem *semaphore.Weighted
+	// ensures only one large file is read at a time.
+	// TODO(nodir): ensure this doesn't hurt performance on SSDs.
+	muLargeFile sync.Mutex
+}
+
+// startProcessing adds the item to the appropriate stage depending on its type.
+func (u *uploader) startProcessing(ctx context.Context, in *UploadInput) error {
+	if in.Path == "" {
+		// Simple case: the blob is in memory.
+		return u.scheduleCheck(ctx, uploadItemFromBlob("", in.Content))
+	}
+
+	// Compute the absolute path only once per directory tree.
+	absPath, err := filepath.Abs(in.Path)
+	if err != nil {
+		return err
+	}
+
+	// Do not use os.Stat() here. We want to know if it is a symlink.
+	info, err := os.Lstat(absPath)
+	if err != nil {
+		return err
+	}
+
+	// visitFile below assumes the file passes the predicate, so evaluate it here.
+	if in.Predicate != nil && !in.Predicate(absPath, info.Mode()) {
+		return nil
+	}
+
+	_, err = u.visitFile(ctx, absPath, info, in.Predicate)
+	return err
+}
+
+// fsCacheKey is the key type for uploader.fsCache.
+type fsCacheKey struct {
+	AbsName      string
+	PredicatePtr uintptr
+}
+
+// fsCacheValue is the value type for uploader.fsCache.
+type fsCacheValue struct {
+	compute  sync.Once
+	dirEntry proto.Message // *repb.FileNode, *repb.DirectoryNode or *repb.SymlinkNode.
```

Comment:
Can we use node for this?
```suggestion
	node proto.Message // *repb.FileNode, *repb.DirectoryNode or *repb.SymlinkNode.
```

Reply:
Done